core/metrics/impl/prometheus/handler_impl.cpp

```diff
@@ -0,0 +1,102 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "handler_impl.hpp"
+#include <prometheus/text_serializer.h>
+#include "log/logger.hpp"
+#include "registry_impl.hpp"
+
+using namespace prometheus;
+
+std::vector<MetricFamily> CollectMetrics(
+    const std::vector<std::weak_ptr<prometheus::Collectable>> &collectables) {
+  auto collected_metrics = std::vector<MetricFamily>{};
+
+  for (auto &&wcollectable : collectables) {
+    auto collectable = wcollectable.lock();
+    if (!collectable) {
+      continue;
+    }
+
+    auto &&metrics = collectable->Collect();
+    collected_metrics.insert(collected_metrics.end(),
+                             std::make_move_iterator(metrics.begin()),
+                             std::make_move_iterator(metrics.end()));
+  }
+
+  return collected_metrics;
+}
+
+namespace kagome::metrics {
+
+  PrometheusHandler::PrometheusHandler()
+      : logger_{log::createLogger("PrometheusHandler", "metrics")} {}
+
+  void PrometheusHandler::onSessionRequest(Session::Request request,
+                                           std::shared_ptr<Session> session) {
+    std::vector<MetricFamily> metrics;
+
+    {
+      std::lock_guard<std::mutex> lock{collectables_mutex_};
+      metrics = CollectMetrics(collectables_);
+    }
+
+    const TextSerializer serializer;
+
+    auto size = writeResponse(session, request, serializer.Serialize(metrics));
+  }
+
+  std::size_t PrometheusHandler::writeResponse(std::shared_ptr<Session> session,
+                                               Session::Request request,
+                                               const std::string &body) {
+    Session::Response res{boost::beast::http::status::ok, request.version()};
+    res.set(boost::beast::http::field::content_type,
+            "text/plain; charset=utf-8");
+    res.content_length(body.size());
+    res.keep_alive(true);
+    res.body() = body;
+    session->respond(res);
+    return body.size();
+  }
+
+  void PrometheusHandler::registerCollectable(Registry *registry) {
+    registerCollectable(
+        dynamic_cast<PrometheusRegistry *>(registry)->registry());
```

Comment:
Please redo this without a dynamic cast. That is going to be **slow**. An inappropriate slowdown in metrics collection mechanism which by design has to be transparent in terms of main code performance affection.

Reply:
I made fixes as were discussed